### PR TITLE
react-query utils proxy

### LINKED ISSUE #2113: Create proxy methods for `react-query`'s `useContext`
We might want to attack https://github.com/trpc/trpc/issues/2124 ahead of this

```ts
function MyComponent() {
  trpc.proxy.post.list.useQuery(/* input */)

  const utils = trpc.proxy.useContext()

  // in a callback:
  utils.post.list.invalidate(/* optional input */)
  utils.post.list.prefetchQuery(/* input */)
  utils.post.list.setQueryData(/* optional input */, /* updater */)

  return <>...</>
}
```

### PR DESCRIPTION
**Pull request by sachinraja**
closes #2113

```ts
const Component = () => {
  const utils = trpc.proxy.useContext()
  
  const deletePostMutation = trpc.proxy.post.delete.useMutation({
    onSuccess(_, { id }) {
      await utils.post.all.invalidate()
    }
  })
}
```
I've omitted `query` and `queries` from the util method names because I think they read better without. The query key is already set for the user, so it's more like you're calling a method off the query, rather than calling a function with the query. This isn't a strong opinion though.

### CODE REVIEW

**Comment by KATT** [comment]
Very coooool

File: examples/next-prisma-todomvc/src/pages/[filter].tsx [hunk 1]
```diff
@@ -70,13 +70,12 @@ function ListItem({ task }: { task: Task }) {
 
   const editTask = trpc.useMutation('todo.edit', {
     async onMutate({ id, data }) {
-      await utils.cancelQuery(['todo.all']);
-      const allTasks = utils.getQueryData(['todo.all']);
+      await utils['todo.all'].cancel();
```
Comment (KATT) [inline examples/next-prisma-todomvc/src/pages/[filter].tsx:73]: I really think this ought to be

```suggestion
      await utils.todo.all.cancel();
```

Any reason why you opted for not doing it that way?

File: packages/react/src/createReactQueryHooksProxy.tsx [hunk 2]
```diff
@@ -107,6 +115,12 @@ export function createReactQueryHooksProxy<
 >(trpc: CreateReactQueryHooks<TRouter, TSSRContext>) {
   const proxy = createProxy((opts) => {
     const args = opts.args;
+
+    if (opts.path[0] === 'useContext') {
+      const context = trpc.useContext();
+      return createReactQueryUtilsProxy(context);
+    }
```
Comment (KATT) [inline packages/react/src/createReactQueryHooksProxy.tsx:122]: Nice

File: packages/react/src/internals/utilsProxy.ts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+};
+
+/**
+ * @internal
+ */
+export type DecoratedProcedureUtilsRecord<TRouter extends AnyRouter> =
+  OmitNeverKeys<{
+    [TKey in keyof TRouter['_def']['record']]: TRouter['_def']['record'][TKey] extends AnyRouter
+      ? DecoratedProcedureUtilsRecord<TRouter['_def']['record'][TKey]>
+      : // utils only apply to queries
```
Comment (KATT) [inline packages/react/src/internals/utilsProxy.ts:153]: Oh, do they? I actually didn't know!

File: packages/react/src/internals/utilsProxy.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+      }
+      case 'getInfiniteData': {
+        const [input] = args;
+        const queryKey = getQueryKey(fullPath, input);
+        return context.getInfiniteQueryData(queryKey);
+      }
+    }
+  });
+  return proxy as DecoratedProcedureUtilsRecord<TRouter>;
+}
```
Comment (KATT) [inline packages/react/src/internals/utilsProxy.ts:233]: This file is impressive, well done 🙏 😻

**Review by KATT — COMMENTED** [review]
Main thing is that we should probably not use the string literal :) Looks awesome

**Comment by KATT** [comment]
Okay, seems to work great! Should we update `/www` too? 🤔

**Comment by sachinraja** [comment]
I think I already updated the relevant docs

**Comment by sachinraja** [comment]
all docs are updated now @KATT

**Pull request merged**